14.0 [MIG] web_copy_confirm

rconjour opened the pull request:
14.0 [MIG] web_copy_confirm

StefanRijnhart committed:
[ADD] Ask for confirmation when copying records

[UPD] Update web_copy_confirm.pot

[ADD] icon.png

rconjour committed:
[IMP] web_copy_confirm: black, isort, prettier

rconjour committed:
[MIG] web_copy_confirm: Migration to 14.0

StefanRijnhart commented on web_copy_confirm/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,6 @@
+=================================================
+Show a confirmation dialogue when copying records
+=================================================
```
You can remove this bit. It is already rendered in the README.rst from the module name. In general, the readme/ fragments are not supposed to contain any RST markup at all.

StefanRijnhart commented on web_copy_confirm/views/assets.xml:
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="utf-8" ?>
-<openerp><data>
+<odoo><data>
```
The `<data>` level in XML files is obsolete.

StefanRijnhart commented:
Looking good in general! I'm confused that a file "setup/web_copy_confirm/odoo_addons/__init__.py" is added in the "pre-commit hook" commit and then removed in the migration commit. Was that a manual action or is the pre-commit hook doing something else by itself?

StefanRijnhart commented:
To my shame I never added a test for this module. Javascript testing is supposed to be easier and more reliable in later versions of Odoo. Would you like to timebox one or two hours to see if you can add a test for the functionality?

rconjour commented:
> Looking good in general! I'm confused that a file "setup/web_copy_confirm/odoo_addons/**init**.py" is added in the "pre-commit hook" commit and then removed in the migration commit. Was that a manual action or is the pre-commit hook doing something else by itself?

I did it in this order:
-  run pre-commit + push (here are the init files created, probably because this is on 8.x)
- migrated the code
- runned pre-commit again (this removes the init file again)

rconjour commented:
> To my shame I never added a test for this module. Javascript testing is supposed to be easier and more reliable in later versions of Odoo. Would you like to timebox one or two hours to see if you can add a test for the functionality?

Sure

StefanRijnhart commented:
> > Looking good in general! I'm confused that a file "setup/web_copy_confirm/odoo_addons/**init**.py" is added in the "pre-commit hook" commit and then removed in the migration commit. Was that a manual action or is the pre-commit hook doing something else by itself?
> 
> I did it in this order:
> 
>     * run pre-commit + push (here are the init files created, probably because this is on 8.x)
> 
>     * migrated the code
> 
>     * runned pre-commit again (this removes the init file again)

Sounds good. I see it is indeed caused by this config: https://github.com/acsone/setuptools-odoo/blob/master/setuptools_odoo/core.py#L33

StefanRijnhart commented on web_copy_confirm/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,2 @@
+This module will show a confirmation dialog when the user selects the
+`Duplicate` option from the `More` dropdown in the standard form view.
```
In 14.0, it's the `Action` dropdown.

rconjour replied:
Changed!

StefanRijnhart commented on web_copy_confirm/readme/USAGE.rst:
```diff
@@ -0,0 +1,2 @@
+This module will show a confirmation dialog when the user selects the
+`Duplicate` option from the `More` dropdown in the standard form view.
```
I think you can leave out this text, as it's the same as the description.

StefanRijnhart commented on web_copy_confirm/readme/HISTORY.rst:
```diff
@@ -0,0 +1,3 @@
+14.0.0.1 (2020-01-04)
```
Align to module version number

StefanRijnhart commented on web_copy_confirm/static/src/js/web_copy_confirm.js:
```diff
@@ -0,0 +1,26 @@
+// Copyright (C) 2018 DynApps <http://www.dynapps.be>
+// @author Stefan Rijnhart <stefan@opener.amsterdam>
```
Please add yourself as an author

StefanRijnhart commented on web_copy_confirm/static/src/js/web_copy_confirm.js:
```diff
@@ -0,0 +1,27 @@
+    const _t = Core._t;
+
+    return FormController.include({
+        _onDuplicateRecord: async function (values, confirmed = false) {
```
I'm a little worried about adding the additional parameter. If another module overrides this method, that could lead to problems because `confirmed` is never passed as true. Is there another way, for instance by letting the dropdown action call a new entry point (for instance, _onDuplicateRecordConfirm)?

rconjour replied:
Agreed, overrided the function in the action list.

PieterPaulussen commented on web_copy_confirm/i18n/web_copy_confirm.pot:
```diff
@@ -0,0 +1,22 @@
+#
+msgid ""
+msgstr ""
+"Project-Id-Version: Odoo Server 8.0\n"
```
Odoo Server 8.0?

rconjour replied:
Regenereated the pot file.

PieterPaulussen commented on web_copy_confirm/static/src/js/web_copy_confirm.js:
```diff
@@ -0,0 +1,45 @@
+    return FormController.include({
+        init: function () {
+            this._super.apply(this, arguments);
+        },
```
Is the override of the init necessary?

rconjour replied:
No, it was not necessary.

StefanRijnhart replied:
Wait, it's not necessary but it allows other modules to override this method so it's good practice to return it.

rconjour replied:
Then, it's back :)

StefanRijnhart replied:
Apologies, my response was to another remark of Pieter's:
```
+    return FormController.include({

The return is also not necessary here I believe?
```

It is this return that I would like to see reintroduced. You can remove the override of the `init` method.

rconjour replied:
Added the return, and removed the init.

PieterPaulussen commented on web_copy_confirm/static/src/js/web_copy_confirm.js:
```diff
@@ -0,0 +1,45 @@
+    const Dialog = require("web.Dialog");
+    const _t = Core._t;
+
+    return FormController.include({
```
The `return` is also not necessary here I believe?

rconjour replied:
Indeed, you were right.

StefanRijnhart replied:
Thanks for putting it back!

PieterPaulussen commented:
I would also like to see a bit more documentation if possible. Odoo's docstrings in javascript have been improved greatly, could you align with the conventions there?

rconjour commented:
> I would also like to see a bit more documentation if possible. Odoo's docstrings in javascript have been improved greatly, could you align with the conventions there?

Added the docblocks.

StefanRijnhart commented on web_copy_confirm/static/src/js/web_copy_confirm.js:
```diff
@@ -0,0 +1,55 @@
+                other_list[duplicate_index].callback = () =>
+                    this._onDuplicateRecordConfirm();
+
+                return Object.assign(props, {
```
You are modifiying `props.items.other` inline so you don't have to reassign it to the toolbar actions object.
![image](https://user-images.githubusercontent.com/1033124/104919327-5d498d80-5996-11eb-8b1e-7d6c4d533c65.png)

StefanRijnhart replied:
Oh, and because of the indentation level, you are not returning the return value of `super` if your condition does not come out as true.

rconjour replied:
Changed!

StefanRijnhart reviewed (approved):
Great work on the tests! Ready for an OCA PR as far as I am concerned.

rconjour commented:
Forcepushed to solve this issue: https://github.com/OCA/web/pull/1791#issuecomment-804685228

Changed:
```
const duplicate_index = other_list.findIndex(
                    (item) =>
                        item.callback.toString() ===
                        "() => this._onDuplicateRecord(this)"
                );
```
to:
```
const duplicate_index = other_list.findIndex(
                    (item) =>
                        item.callback.toString().replace(/\s/g, "") ===
                        "()=>this._onDuplicateRecord(this)"
                );
```

And added an check if the index really exists:
```
if (other_list[duplicate_index]) {
                    other_list[duplicate_index].callback = () =>
                        this._onDuplicateRecordConfirm();
                }
```

StefanRijnhart commented:
Maybe it's more reliable to fetch the item by the `description` (`_t("Duplicate"`) rather than the string representation of the `callback`?

StefanRijnhart commented:
OK, now working with two PRs is getting confusing. Let's continue the conversation on the public PR
